### ISSUE TITLE
Feature: Improve elevation of WSL commands

### ISSUE THREAD

**Issue by felipecrs** [gerardog/gsudo#55]
It works in PowerShell,  but when used in bash, it opens a new shell not on bash anymore. See:

![yan7KMwO7d](https://user-images.githubusercontent.com/29582865/89743182-e9b26480-da76-11ea-8044-5d29c3aca42c.gif)

**Comment by gerardog** [comment]
Well I considered it once, what stopped me was how to handle when you have more than one wsl distro. Are you aware that you can `gsudo.exe wsl` or `gsudo wsl -d Alpine` and that will keep you on wsl? (And you can create an alias for that) Ans you looking for an easy way, just out-of-the-box ?

**Comment by felipecassiors** [comment]
On WSL, there is an environment variable that might be useful: 

```sh-session
$ echo $WSL_DISTRO_NAME
Ubuntu
```

**Comment by felipecassiors** [comment]
> Well I considered it once, what stopped me was how to handle when you have more than one wsl distro. Are you aware that you can `gsudo.exe wsl` or `gsudo wsl -d Alpine` and that will keep you on wsl? (And you can create an alias for that) Are you looking for an easy way, just out-of-the-box ?

That's cool! I'll create an alias for that. It would be also cool if it was supported out-of-the-box.

**Comment by gerardog** [comment]
> ```
> $ echo $WSL_DISTRO_NAME
> Ubuntu
> ```

I would need to first set `set WSLENV=WSL_DISTRO_NAME/w` in order to read that env var from a c# Process... 🤔
Maybe I could find the parent 'wsl.exe' process and read it's command line arguments...

**Comment by felipecassiors** [comment]
> > ```
> > $ echo $WSL_DISTRO_NAME
> > Ubuntu
> > ```
> 
> I would need to first set `set WSLENV=WSL_DISTRO_NAME/w` in order to read that env var from a c# Process... 🤔
> Maybe I could find the parent 'wsl.exe' process and read it's command line arguments...

Finding the parent `wsl.exe` command-line call would be better indeed.

**Comment by gerardog** [comment]
Maybe you can try this:

go to `C:\ProgramData\chocolatey\lib\gsudo\bin` if you installed via choco,
or  `C:\Users\ggrignoli\scoop\apps\gsudo\current` if used scoop, 

then create a file named `gsudo`  (no extension, don't use sudo because it will collision with linux sudo)
```
#!/bin/sh
gsudo.exe wsl -d $WSL_DISTRO_NAME -e $*
```

Please note with this file, `gsudo command1` will run `command1` using WSL, so requires a bash syntax. like 'gsudo cat /proc/cpuinfo' 
As opposed to: `gsudo.exe command1` that will actually do `cmd.exe /c command1`.
(Is that weird? Any better ideas?)

If you do, could you please let me know how it worked? 
I could just ship this file in the next release. I am interested to hear what usability would be better.

**Comment by felipecassiors** [comment]
After fixing `shellcheck` recommendations:
```sh
#!/bin/sh
gsudo.exe wsl -d "$WSL_DISTRO_NAME" -e "$@"

```

And it works to elevate the current shell and for `gsudo cat /proc/cpuinfo`, but does not work for running other commands like:

````sh-session
$ gsudo cup.exe all
Info: Process exited with code 1
```

**Comment by felipecassiors** [comment]
> (Is that weird? Any better ideas?)

I don't think it's weird. Inside of bash, it feels more natural to use bash syntax and not CMD. This means that if you want to run a Windows command, you must use the `.exe`.

**Comment by gerardog** [comment]
> And it works to elevate the current shell and for gsudo cat /proc/cpuinfo, but does not work for running other commands like:
>```
> $ gsudo cup.exe all
> Info: Process exited with code 1
>```

Oh, it's weird that `wsl -e notepad.exe` fails, but `notepad.exe` from bash works... to fix it the script has to detect if you are trying to jump to the windows world... parsing the command is unlikely to reliable or nice code... :(

**Comment by felipecassiors** [comment]
The solution is to do not use `-e` at all:

```bash
#!/bin/sh
gsudo.exe wsl -d "$WSL_DISTRO_NAME" -- "$@"
```

**Comment by felipecassiors** [comment]
It works as intended now:

![image](https://user-images.githubusercontent.com/29582865/89797661-90434780-db01-11ea-8dd7-4319d3b47a19.png)

**Comment by felipecassiors** [comment]
The downside is that it will always start the process using the default shell, even if you're not on that shell at that time. So, if you're in bash and type gsudo, you will end up with a zsh shell.

**Comment by gerardog** [comment]
> However, are you sure about writing Info: Process exited with code X in the stdout? I don't think it's a good idea, except forwarding the status code number itself.

I thought it was a good idea back when I was the only one using it. I guess it doesn't make sense to display it (other than showing off 'look dude, you can catch the code' 🤣). I will consider turning it into a DEBUG loglevel.  In the meantime try `gsudo.exe config LogLevel Warning` to narrow the verbosity.

**Comment by gerardog** [comment]
> The downside is that it will always start the process using the default shell, even if you're not on that shell at that time. So, if you're in bash and type gsudo, you will end up with a zsh shell.

What about the $SHELL variable? I've tried but the `gsudo notepad.exe` scenario broke again :(

**Comment by felipecassiors** [comment]
SHELL does not point to the current shell, it points to the default shell. I ended up with:

```bash
#!/bin/bash -x

set -e

# https://askubuntu.com/a/1022440/1074320
shell=("$(sh -c 'ps -p $$ -o ppid=' | xargs -I'{}' readlink -f '/proc/{}/exe')")

if [ $# -gt 0 ]; then
    shell+=("-c")
fi

exec gsudo.exe wsl -d "$WSL_DISTRO_NAME" -e "${shell[@]}" "$*"
```

**Comment by felipecassiors** [comment]
It's not possible to determine the parent shell which was used to call a script from a script. I could only make it work by using this function:

```bash
function gsudo() {
    local command_line
    command_line=("$(sh -c 'ps -p $$ -o ppid=' | xargs -I'{}' readlink -f '/proc/{}/exe')")
    if [ $# -gt 0 ]; then
        command_line+=("-c" "$*")
    fi
    gsudo.exe wsl -d "$WSL_DISTRO_NAME" -e "${command_line[@]}"
}
```

I put it into my `.bash_aliases` because I run this file on both `bash` and `zsh` initialization.

**Comment by gerardog** [comment]
Thanks for sharing your progress and final solution. I don't think gsudo installer should edit the `.bash_aliases` of every installed wsl distro. But I can use a variant of your previous version: If I can't figure out the current shell, then fallback to the default `$SHELL` is not a so bad compromise, I think, for most people.

**Comment by felipecassiors** [comment]
In that case, you don't need anything but:

```bash
#!/bin/sh
gsudo.exe wsl -d "$WSL_DISTRO_NAME" -- "$@"
```

No need to check `SHELL` variable, `wsl` takes care of it to you if not using `-e`.

**Comment by felipecrs** [comment]
The `asciinema`, `chezmoi`, they all start new shell instances based on the default shell and not on the currently running shell, so I believe it worth including such a shim on Gsudo installs without worrying about this.

**Comment by gerardog** [comment]
Yes. Also Unix/Linux `su` just uses the default shell. 
Users can still change the default shell, or wrap `gsudo`/`gsudo.exe` into custom scripts.

**Comment by gerardog** [comment]
Just a small follow up and note to self: This new bash script is executed when the user runs `gsudo` from `git-bash` for windows (and it breaks). So that scenario needs to be tested too.

**Comment by gerardog** [comment]
If the bash script is named `gsudo` (no extension) it is picked up by WSL, git-bash and powershell. But powershell just throws an error. So I'll name it `gsudo.sh` instead which seems to work so far.

**Comment by felipecrs** [comment]
I wonder what is the recommended approach for including a binary in WSL through a Windows application. 

Docker Desktop does that, by the way.

**Comment by gerardog** [comment]
Fixed in v1.0 #94

**Comment by gerardog** [comment]
Implemented in [v1.0.0](https://github.com/gerardog/gsudo/releases/tag/v1.0.0) (PR #94)
